src/apscheduler/datastores/asyncpg.py

```diff
@@ -0,0 +1,582 @@
+        def after_attempt(self, retry_state: tenacity.RetryCallState) -> None:
+            self._logger.warning(
+                "Temporary data store error (attempt %d): %s",
+                retry_state.attempt_number,
+                retry_state.outcome.exception(),
+            )
+
+        # OSError is raised by asyncpg if it can't connect
+        return tenacity.AsyncRetrying(
+            stop=self.retry_settings.stop,
+            wait=self.retry_settings.wait,
+            retry=tenacity.retry_if_exception_type(OSError),
+            after=after_attempt,
+            sleep=anyio.sleep,
+            reraise=True,
+        )
+
+    async def _create_schema(self, conn: Connection):
+        with open(Path(__file__).parent / "asyncpg.sql") as f:
+            query = f.read()
+        async with conn.transaction():
+            await conn.execute(f"create schema if not exists {self.schema}")
+            await conn.execute(query)
+
+    async def start(
+        self, exit_stack: AsyncExitStack, event_broker: EventBroker
+    ) -> None:
+        await super().start(exit_stack, event_broker)
+        asynclib = sniffio.current_async_library() or "(unknown)"
+        if asynclib != "asyncio":
+            raise RuntimeError(
+                f"This data store requires asyncio; currently running: {asynclib}"
+            )
+
+        # Verify that the schema is in place
+        async for attempt in self._retry():
+            with attempt:
+                async with self.pool.acquire() as conn:
+                    if self.start_from_scratch:
+                        for table_name in self._tables:
+                            await conn.execute(f"drop table if exists {table_name}")
+                        for dbtype in self._dbtypes:
+                            await conn.execute(f"drop type if exists {dbtype}")
+                    await self._create_schema(conn)
+                    version = await conn.fetchval("select schema_version from metadata")
+                    if version is None:
+                        r = await conn.execute(
+                            "insert into metadata(schema_version) values (1)"
+                        )
+                        print(f"{table_name}: {r}")
+                    elif version > 1:
+                        raise RuntimeError(
+                            f"Unexpected schema version ({version}); "
+                            f"only version 1 is supported by this version of "
+                            f"APScheduler"
+                        )
+
+    async def add_task(self, task: Task) -> None:
+        query = """insert into tasks (id, func, executor, max_running_jobs, misfire_grace_time)
+        values ($1, $2, $3, $4, $5)
+        on conflict (id) do update set func=$2, executor=$3, max_running_jobs=$4, misfire_grace_time=$5
+        where tasks.id=$1
+        returning case when xmax::text::int > 0 then 'updated' else 'inserted' end
```

Comment:
I find this rather elegant compared to the sqlalchemy version, but again not sure it's portable since we have the comfort here knowing what db we're dealing with